### PR TITLE
Add reference-counted versions of `RwLock` guards

### LINKED ISSUE #33: Add owned variants of RwLock guards
There exists a method to lock a `Mutex` in an `Arc` using `lock_arc`, so it would make sense for there to be `read_arc`/`write_arc` methods for the `RwLock`.

Related: async-rs/async-std#991

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
Fixes #33

### CODE REVIEW

**Review by notgull — CHANGES_REQUESTED** [review]
The `rwlock.rs` file is getting to be a bit long; I think it should be split into multiple files. As a rule of thumb, a single module shouldn't be more than 1000 lines.

(I'm busy for the time being and won't be able to take a more in depth look until the weekend).

File: src/rwlock.rs [hunk 1]
```diff
@@ -304,161 +501,143 @@ impl<T: fmt::Debug + ?Sized> fmt::Debug for RwLock<T> {
...
-/// The future returned by [`RwLock::write`].
-pub struct Write<'a, T: ?Sized> {
-    /// Raw write lock acquisition future, doesn't depend on `T`.
-    raw: RawWrite<'a>,
-
-    /// Pointer to the value protected by the lock. Invariant in `T`.
-    value: *mut T,
+/// An owned, reference-counting guard that releases the read lock when dropped.
+#[clippy::has_significant_drop]
+pub struct ArcRwLockReadGuard<T> {
```
Comment (notgull) [inline src/rwlock.rs:566]: Could you change this to `RwLockReadGuardArc` for consistency with the rest of the crate?

File: src/rwlock.rs [hunk 2]
```diff
@@ -304,161 +501,143 @@ impl<T: fmt::Debug + ?Sized> fmt::Debug for RwLock<T> {
...
-        f.write_str("Write { .. }")
+impl<T> ArcRwLockReadGuard<T> {
+    /// Constructs the underlying `Arc` back from the underlying `RwLock`.
+    ///
+    /// # Safety
+    ///
+    /// Both the returned `Arc` and the guard will decrement their reference
+    /// counts on drop! So one of the two must be forgotten.
+    #[inline]
+    unsafe fn inner_arc(guard: &Self) -> Arc<RwLock<T>> {
```
Comment (notgull) [inline src/rwlock.rs:584]: It might be less error-prone to return an `ManuallyDrop<Arc<RwLock<T>>>` instead.

File: src/rwlock.rs [hunk 3]
```diff
@@ -595,12 +774,14 @@ impl<'a, T: ?Sized> RwLockUpgradableReadGuard<'a, T> {
...
 
 impl<T: fmt::Debug + ?Sized> fmt::Debug for RwLockUpgradableReadGuard<'_, T> {
+    #[inline]
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         fmt::Debug::fmt(&**self, f)
     }
 }
 
 impl<T: fmt::Display + ?Sized> fmt::Display for RwLockUpgradableReadGuard<'_, T> {
+    #[inline]
```
Comment (notgull) [inline src/rwlock.rs:784]: Is there any reason why you've inlined several methods here?

File: src/rwlock/futures.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    /// `'static` is a lie, this field is actually referencing the
+    /// `Arc` data. But since this struct also stores said `Arc`, we know
+    /// this value will be alive as long as the struct is.
+    ///
+    /// Yes, one field of the `ArcUpgrade` struct is referencing another.
+    /// Such self-references are usually not sound without pinning.
+    /// However, in this case, there is an indirection via the heap;
+    /// moving the `ArcUpgrade` won't move the heap allocation of the `Arc`,
+    /// so the reference inside `RawUpgrade` isn't invalidated.
+    pub(super) raw: RawUpgrade<'static>,
```
Comment (notgull) [inline src/rwlock/futures.rs:262]: Overall this seems sound, but this field in particular should be wrapped in a `ManuallyDrop` and explicitly dropped before the `lock` field is. That way there's no chance for a dangling reference.

File: src/rwlock.rs (hunk 3 above)
Comment (Jules-Bertholet) [reply src/rwlock.rs:784]: Consistency with the equivalent trivial methods on the `Arc` guards. I can remove them if you like

**Review by notgull — APPROVED** [review]
Thanks!

**Pull request merged**